File: content/private/_holding/LeetCode/03_Sliding_Window/121-best-time-to-buy-and-sell-stock.cpp
// LeetCode 121 — Best Time to Buy and Sell Stock (C++). Contraste con .py.
// g++ -std=c++20 -Wall -Wextra -fsanitize=address 121-best-time-to-buy-and-sell-stock.cpp -o s && ./s
#include <algorithm>
#include <cassert>
#include <climits>
#include <vector>

class Solution {
 public:
  int maxProfit(std::vector<int>& prices) {
    int min_price = INT_MAX;  // Python: float('inf')
    int best = 0;
    for (int p : prices) {
      min_price = std::min(min_price, p);
      best = std::max(best, p - min_price);
    }
    return best;
  }
};

int main() {
  Solution s;
  std::vector<int> a = {7, 1, 5, 3, 6, 4};
  std::vector<int> b = {7, 6, 4, 3, 1};
  assert(s.maxProfit(a) == 5);
  assert(s.maxProfit(b) == 0);
  return 0;
}

File: content/private/_holding/LeetCode/03_Sliding_Window/121-best-time-to-buy-and-sell-stock.py
## Solución 1- Fuerza bruta

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        best = 0
        for i in range(n):
            for j in range(n+1,n):
                best = max(best, prices[j] - prices[i])
            return best
        
## Tiempo - O(n²) porque hay dos bucles
## Espacio - O(1)

# Solución 2 - One-pass tracking del mínimo

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float("inf")
        best = 0
        for price in prices:
            min_price = min(min_price, price)
            best = max(best, price - min_price)
        return best
    
# Tiempo - O(n)
# Espacio - O(1)

# Solución 3 - Two Pointers / sliding window explícito

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left, right = 0, 1
        best = 0
        while right < len(prices):
            if prices[left] < prices[right]:
                best = max(best, prices[right] - prices[left])
            else:
                left = right
            right += 1
        return best
    
    